File: 169.majority-element.go
/*
 * @lc app=leetcode id=169 lang=golang
 *
 * [169] Majority Element
 */

// @lc code=start
func majorityElement(nums []int) int {
	res, count := nums[0], 0

	for _, num := range nums {
		if count == 0 {
			res, count = num, 1
		} else {
			if num == res {
				count++
			} else {
				count--
			}
		}
	}

	return res
}
// @lc code=end

File: 169.majority-element.cpp
/*
 * @lc app=leetcode id=169 lang=cpp
 *
 * [169] Majority Element
 */

// @lc code=start
class Solution {
 public:
  int majorityElement(vector<int>& nums) {
    int maj = INT_MIN;
    int cnt = 0;

    for (int i = 0; i < nums.size(); ++i) {
      if (cnt == 0) maj = nums[i];

      if (nums[i] == maj)
        cnt++;
      else
        cnt--;
    }

    return maj;
  }
};
// @lc code=end

